app/questionnaire/navigation.py

```diff
@@ -31,110 +31,133 @@ def build_navigation(self, current_group_id, current_group_instance):
+                return block['id']
+
+        return last_block
+
+    def _can_reach_summary_confirmation(self, last_section_group, last_block_id):
+        if self.routing_path and Location(last_section_group, 0, last_block_id) in self.routing_path and \
+                len(self.completed_blocks) > 0 and set(self.routing_path[:-1]).issubset(self.completed_blocks):
+            return True
+
+        return False
+
+    def _build_single_navigation(self, section, current_group_id, first_location, last_block):
+        default_highlight = current_group_id in section['group_order']
+        is_completed = self._is_completed(last_block)
 
-    def _build_repeating_navigation(self, repeating_rule, group, current_group_id, current_group_instance):
+        return self._generate_item(section['title'], is_completed, first_location, default_highlight)
+
+    def _build_repeating_navigation(self, repeating_rule, section, current_group_id,
+                                    current_group_instance, completed_id):
+        group = SchemaHelper.get_group(self.survey_json, section['group_order'][0])
         first_location = Location(group['id'], 0, group['blocks'][0]['id'])
         no_of_repeats = evaluate_repeat(repeating_rule, self.answer_store)
 
         repeating_nav = []
 
         if repeating_rule['type'] == 'answer_count':
-            completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
             is_current_group = group['id'] == current_group_id
-            link_names = self._generate_link_names(repeating_rule)
+            link_names = self._generate_link_names(section['title_from_answers'])
 
             repeating_nav = self._generate_repeated_items(link_names, first_location, completed_id, no_of_repeats,
                                                           is_current_group, current_group_instance)
 
         elif no_of_repeats > 0:
-            item = self._build_single_navigation(group, current_group_id, first_location)
+            item = self._build_single_repeating_navigation(section, current_group_id, first_location,
+                                                           completed_id, no_of_repeats)
             repeating_nav.append(item)
 
         return repeating_nav
 
-    def _build_single_navigation(self, group, current_group_id, first_location):
-        completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
-        default_highlight = self._should_highlight(group, current_group_id)
-        is_completed = self._is_complete_non_repeating(completed_id)
-
-        return self._generate_item(group['title'], is_completed, first_location, default_highlight)
+    def _build_single_repeating_navigation(self, section, current_group_id, first_location,
+                                           completed_id, no_of_repeats):
+        default_highlight = current_group_id in section['group_order']
```

Comment:
Not sure why this was called `default_highlight` previously, it should be renamed something like `should_highlight`